session_4/blogspot/README.md

```diff
@@ -1,54 +1,239 @@
+  .
+  .
+  def show
+   @pagy, @articles = pagy(Article.order("views DESC").all)
+   .
+   .
+  end
+  
+```
+
+Also this needs to be meantion inside `ArticlesHelper` module
+
+```ruby
+  module ArticlesHelper
+    include Pagy::Frontend
+  end
+```
+
+And inside `articles\index.html.erb` we use these variables as shown also with integration with bootstrap pagination we have everyhing on track
+
+```erb
+  <%#Normal articles loops%>
+  <% @articles.each do |article| %>
+   .
+   .
+  <% end %>
+  .
+  .
+  <%#Those "==" is same as "= raw pagy_bootstrap_nav(@pagy).html_safe"%>
+  <%== pagy_bootstrap_nav(@pagy).html_safe%>
+```
+
+The other Feature is to count the number of views, also this field is useful in sorting the article (shown later)
+
+<img src="screenshot/11.PNG" height = "250">
+
+These views gets added up once the blog is been visited(show)
+
+``` ruby
+ def show
+ ..
+  if @article.views
+        views = @article.views + 1
+        @article.views = views
+      else
+        @article.views = 1
+  end
+  ..
+ end
+```
+
+# Adding Several Posts
+ Now to demonstrate Pagination feature and sorting, 3 more new articles are been added
+ 
+<img src="screenshot/12.PNG" height = "250">
+
+ As seen in these two image the articles are been arranged in decending order
+ 
+ ``` ruby
+    Article.order("views DESC").all
+ ```
+Also this Pagy gem uses `AJAX` request which present the indices to users without any reload, this makes the browsing experience more smoother.
+
+# Live Toast Message
+ Now seen is a ground breaking, more over head breaking feature. This gave users the feature to recive a toast when ever a new blog has been posted.
+ So decided to implement this feature without using any gem.
+ 
+ Making a new sesion
+ 
+ <img src="screenshot/14.PNG" height = "250">
+ 
+ LoggingIn to a Test User
+ 
+ <img src="screenshot/15.PNG" height = "250">
+ 
+ Making an Article
+ 
+ <img src="screenshot/16.PNG" height = "250">
+ 
+ After Publishing, within a second there will be a toast message to others with prior fields.
+ 
+ <img src="screenshot/13.PNG" height = "250"> <img src="screenshot/15.Popup on other window.PNG" height = "250">
+ 
+ The screen for the Test User without any popup which is been noted, also there exists a problem which couldn't Figured Out! To update the index page when new article added.
+ This is because of the gem usage, also those index get updated from the partial by the gem. But anyways the users recieve popup messages.
+ 
+ <img src="screenshot/16. No Popup on other session.PNG" height = "250">
+ 
+ So now lets see who is get implemented. First to get a regular update from the rails server. For this to happen we need a simple get server to run from client(user) side.
+ Implementation is simple we run a javascript which run `GET` requests over a special route(explained later), this route in turn return a data in `JSON`. Also to be noted
+ this request runs `AJAX` so the page wont get refreshed.
+ 
+ ``` javascript
```

Comment:
Alright, this is my favorite part of your website. 

What you have implemented is called "[long polling](https://www.pubnub.com/blog/http-long-polling/)" - which is a nice first step. But long polling is _not efficient_: every time you make an AJAX request and there are no new articles, it's wasting the time and resources of your server.

Alternatively, you can configure the server such that it notifies the client when a new article is published- meaning that client does not have to make a request every 1000 milliseconds. You are using resources only _when_ a new article is created, which is more efficient. 

The alternative method is called [WebSockets](https://www.ably.io/blog/websockets-vs-long-polling) and you can read up on how to [implement WebSockets in Rails](https://edgeguides.rubyonrails.org/action_cable_overview.html).

Reply:
Woah!! I was searching for this exact thing.
>But long polling is not efficient: every time you make an AJAX request and there are no new articles, it's wasting the time and resources of your server

You were right my virtual machine was slowing down when i tried to open many sessions. Noted!!